### PR TITLE
feat: Support Follow Scheme Requests

### PR DESCRIPTION
**Pull request by qiqizjl**
### What this PR does / why we need it:

Support Follow Scheme Requests #5041 
### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Review by spacewander — COMMENTED** [review]
I am afraid this can't be done as easily as this.
APISIX uses the upstream as a cache key in many places. So an upstream can't have undetermined ports. That's why I add a check here: https://github.com/apache/apisix/blob/eea5cd02fc9209eee9d169bbb0bb1d45678226ce/apisix/upstream.lua#L171

File: apisix/schema_def.lua [hunk 1]
```diff
@@ -438,7 +438,7 @@ local upstream_schema = {
         },
         scheme = {
             default = "http",
-            enum = {"grpc", "grpcs", "http", "https"}
+            enum = {"grpc", "grpcs", "http", "https","http_auto"}
```
Comment (arthur-zhang) [inline apisix/schema_def.lua:441]: space here

File: docs/zh/latest/admin-api.md [hunk 2]
```diff
@@ -560,7 +560,7 @@ APISIX 的 Upstream 除了基本的负载均衡算法选择外，还支持对上
...
-| scheme         | 可选                               | 辅助           | 跟上游通信时使用的 scheme。需要是 ['http', 'https', 'grpc', 'grpcs'] 其中的一个，默认是 'http'。                                                                                                                                                                                                                                                            |
+| scheme         | 可选                               | 辅助           | 跟上游通信时使用的 scheme。需要是 ['http', 'https', 'http_auto' ,'grpc', 'grpcs'] 其中的一个，默认是 'http'。 `http_auto`:跟随请求的协议与上游通讯                                                                                                                                                                        |
```
Comment (arthur-zhang) [inline docs/zh/latest/admin-api.md:563]: `'http_auto', 'grpc'`

File: t/node/upstream-http-auto.t [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+# limitations under the License.
+#
+use t::APISIX 'no_plan';
+
+repeat_each(1);
+#no_long_string();
+no_root_location();
+log_level('info');
+run_tests;
+
```
Comment (arthur-zhang) [inline t/node/upstream-http-auto.t:24]: try add block preprocessor to simplify test code

```
add_block_preprocessor(sub {
    my ($block) = @_;

    if (!$block->request) {
        $block->set_value("request", "GET /t");
    }

    if (!$block->no_error_log && !$block->error_log) {
        $block->set_value("no_error_log", "[error]\n[alert]");
    }
});
```

File: apisix/balancer.lua [hunk 4]
```diff
@@ -328,6 +328,12 @@ function _M.run(route, ctx, plugin_funcs)
         end
     end
 
+    -- scheme is http_auto, need dynamic get port
+    local scheme = ctx.matched_upstream.scheme  or "http"
+    if scheme == "http_auto" and (server.port == 80 and server.prot == 443) then
```
Comment (arthur-zhang) [inline apisix/balancer.lua:333]: and to or?

Comment (arthur-zhang) [inline apisix/balancer.lua:333]: prot ?

is it should be `server.port == 80 or server.port == 443`

**Comment by moonming** [comment]
@qiqizjl Do you need any help? Looking forward to your first PR merged

**Pull request closed without merging**